modules/containers/include/hephaestus/containers/fixed_circular_buffer.h

```diff
@@ -0,0 +1,133 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <array>
+#include <concepts>
+#include <cstddef>
+#include <optional>
+#include <utility>
+
+namespace heph::containers {
+/// Implementation of a fixed capacity circular buffer
+///
+/// \tparam T the value type of the elements
+/// \tparam Capacity Maximum capacity. Should be a power of two for efficiency reasons.
+///
+/// \note This class is not thread safe. If you need to push/pop from differeent threads, protect it with a
+/// lock!
+template <typename T, std::size_t Capacity>
+class FixedCircularBuffer {
+public:
+  static_assert((Capacity & (Capacity - 1)) == 0, "Capacity should be a power of two.");
+
+  FixedCircularBuffer() = default;
+  ~FixedCircularBuffer() = default;
+  FixedCircularBuffer(FixedCircularBuffer&& other) noexcept;
+  auto operator=(FixedCircularBuffer&& other) noexcept -> FixedCircularBuffer&;
+
+  FixedCircularBuffer(const FixedCircularBuffer& other) = default;
+  auto operator=(const FixedCircularBuffer& other) -> FixedCircularBuffer& = default;
+
+  [[nodiscard]] auto empty() const -> bool;
+  [[nodiscard]] auto full() const -> bool;
+  [[nodiscard]] auto size() const -> std::size_t;
+
+  /// Pushes \param u onto the buffer only if the buffer is not full.
+  ///
+  /// \return true if space was available, false otherwise
+  template <std::convertible_to<T> U>
+  auto push(U&& u) -> bool;
```

Comment:
```suggestion
  [[nodiscard]] auto push(U&& u) -> bool;
```

The user should think about what happens when the queue is full.